README.md
hadupils
========

Operating environment oriented utilities for hadoop (Hadoop + Utils => hadupils)

## Shell Environment Variables
- $HADUPILS_TMP_PATH
    * This is the base path for DFS temporary file/directory creation
    * Defaults to '/tmp' on the DFS (only set this if you need another base directory)
    * Command 'cleanup' will use this ENV var for the base tmp_path to look for /hadupils-tmp*/
      tmpdirs if the tmp_path isn't set throught the command line
    * Other commands that use this are: mktemp, withtmpdir

- $HADUPILS_TMPDIR_PATH
    * Set when the subcommand is executed in a subshell via the hadupils 'withtmpdir' command
    * The value comes from the tmp directory that hadupils created for the subcommand
    * It will cleanup (remove) the directory if the subcommand returns an exitstatus of zero
- $HADUPILS_TMP_TTL
    * This is the Time-To-Live for hadupils DFS temporary files/directories (hadupils-tmp*)
    * Defaults to '1209600' (2 weeks)
    * Command 'cleanup' will use this ENV var to remove any /hadupils-tmp*/ tmpdirs within
      $HADUPILS_TMP_PATH where all files within are older than TTL, (Time.now.utc - $HADUPILS_TMP_TTL)
      if ttl isn't set through the command line

## Hadpuils' Commands
- hive __command__ _options_
- hadoop __command__ _options_
- mktemp [-d]
- withtmpdir __subshell_command__
- rm [-rR] __full_path_to_file_or_directory__
- cleanup [-n] __full_path_to_tmp_dir__ __ttl__

### Example Usages
``` shell
hadupils hive -e 'select a.col from tab1 a'
hadupils hadoop fs -ls /tmp
hadupils mktemp -d
hadupils withtmpdir 'echo $HADUPILS_TMPDIR_PATH'
hadupils rm -r /tmp/hadupils-tmp-e341afe01721013128c122000af92329
hadupils cleanup -n
```


CHANGELOG.md

### 0.1.0

* Basic functionality for representing a command, extensions,
  assets, runners.
* A hive runner enforcing user config and hadoop-ext extension
* A hadupils executable for entering the command model

### 0.1.1

* Removed evil rubygems requirement from executable
* Added this glorious changelog

### 0.1.2

* Fixed embarrassing mispelling of "shoulda-context" in gemspec
  development dependencies

### 0.1.3

* Fixed compat. issue for ruby 1.8; downcase on symbol is a no-no.
  In Hadupils::Command module.

### 0.2.0

* Introduced hive extensions (hive-ext)
* Hive command uses hive extensions to assemble hivercs
* Hive command assembles HIVE_AUX_JARS_PATH variable
* Base runner can handle environment variable hash as first command
  parameter (custom support for pre-1.9 ruby, the Kernel.system call
  of which does not handle such things)

### 0.3.0

* Introduced Hadupils::Extensions::FlatArchivePath
* The "hive" command uses a FlatArchivePath for hadoop-ext instead
  of Flat, so PATH will include bin dirs of any hadoop-ext archives
  when a streaming query runs.
* Some misc. utility functions in Hadupils::Util for reading tarballs.

### 0.4.0

* Introduced Hadupils::Extensions::Hive.build_archive
  Helper method for assembling gzipped archives the contents of which
  are hive-ext compatible.

### 0.5.0

* Introduced a hadoop command enforcing user config and hadoop-ext
  configuration files (hadoop.xml) that calls the hadoop runner
* Introduced the following commands and runners that utilize the hadoop
  runner (to work against the DFS): mktemp, withtmpdir and rm
* Support for hadoop_confs in Hadupils::Extensions
* Introduced Hadupils::Extensions::Dfs::TmpFile
* Introduced Hadupils::Hacks module for String Refinements (self.randcase)
  for Ruby 2+ and Monkey Patching for the String class for Ruby < 2.0
* Introduced $HADUPILS_BASE_TMP_PATH and $HADUPILS_TMPDIR_PATH for use with
  commands: mktemp, withtmpdir and rm
* Some refactoring and fixed a bug with the specs for Mac OS X
* Tweaked old unit tests and added new ones for the new features
* Updated the README with examples

### 0.6.0

* Renamed $HADUPILS_BASE_TMP_PATH to $HADUPILS_TMP_PATH (less typing)
* Introduced $HADUPILS_TMP_TTL for use with command: cleanup
* Introduced Hadupils::Commands::Cleanup to identify and remove old hadupils tmp DFS
  directories/files where all files within any hadupils-tmpdir* in $HADUPILS_TMP_PATH
  are older than $HADUPILS_TMP_TTL, the TTL (Time.now.utc - $HADUPILS_TMP_TTL)
* The Hadupils::Runnders::Base.new.execute! method now uses Open3.capture2 or Kernel.system
* Fixed 1.8.7 compatibility bug with the Kernel.system call in
  Hadupils::Extensions::Hive::AuxJarsPath.build_archive
* Some refactoring

### 0.6.1

* Fixed weird bug in Hadupils::Extensions::Hive.build_archive; the call
  was hanging reading from stderr, which isn't strictly necessary anyway.
  Now it no hangy.

### 0.6.2
* Stopped the hadupils cleanup command from outputting 'Removing...' to STDOUT when it
  doesn't have anything to actually remove.

### 0.7.0
* Now using Process.spawn instead of Kernel.system, the ruby way for 1.9+
* Fixed subcommand process deadlocking when any IO pipe buffer became full (Ruby 1.9+) via
  threads and process waiting, the ruby way or a ruby way at least ;)
* $HADUPILS_TMP_TTL now defaults to 1209600 (last 2 weeks)
* Passing unit tests

### 0.7.1
* Cleanup command now outputs additional context, candidate tmpdir, to help identify the
  runtime state for long running cleanups
* Now ignoring the 'vendor' directory in .gitignore

### 0.7.2
* $HADUPILS_TMP_TTL default reverts back to 86400 (last 24 hours)


lib/hadupils/extensions.rb
require 'uuid'
require 'open3'
require 'tempfile'

module Hadupils::Extensions
  # Tools for managing shell commands/output performed by the runners
  module Runners
    module Shell
      def self.command(*command_list)
        opts = {}

        begin
          if RUBY_VERSION < '1.9'
            Open3.popen3(*command_list) do |i, o, e|
              @stdout = o.read
              @stderr = e.read
            end
            @status = $?
            $stdout.puts @stdout unless @stdout.nil? || @stdout.empty? || Shell.silence_stdout?
            $stderr.puts @stderr unless @stderr.nil? || @stderr.empty?
            @stdout = nil unless capture_stdout?
            @stderr = nil unless capture_stderr?
          else
            stdout_rd, stdout_wr  = IO.pipe     if capture_stdout?
            stderr_rd, stderr_wr  = IO.pipe     if capture_stderr?
            opts[:out]            = stdout_wr   if capture_stdout?
            opts[:err]            = stderr_wr   if capture_stderr?

            # NOTE: eval prevents Ruby 1.8.7 from throwing a syntax error on Ruby 1.9+ syntax
            pid = eval 'Process.spawn(*command_list, opts)'
            @status = :waiting unless pid.nil?

            Thread.new do
              Process.wait pid
              @status = $?
            end

            Thread.new do
              if capture_stdout?
                stdout_wr.close
                @stdout = stdout_rd.read
                stdout_rd.close
                $stdout.puts @stdout unless @stdout.nil? || @stdout.empty? || Shell.silence_stdout?
              end
            end

            Thread.new do
              if capture_stderr?
                stderr_wr.close
                @stderr = stderr_rd.read
                stderr_rd.close
                $stderr.puts @stderr unless @stderr.nil? || @stderr.empty?
              end
            end
            sleep 0.1 while @status == :waiting
          end

          [@stdout, @stderr, @status]
        rescue Errno::ENOENT => e
          $stderr.puts e
          [@stdout, @stderr, nil]
        end
      end

      def self.capture_stderr?
        @capture_stderr
      end

      def self.capture_stderr=(value)
        @capture_stderr = value
      end

      def self.capture_stdout?
        @capture_stdout || Shell.silence_stdout?
      end

      def self.capture_stdout=(value)
        @capture_stdout = value
      end

      def self.silence_stdout?
        @silence_stdout
      end

      def self.silence_stdout=(value)
        @silence_stdout = value
      end
    end
  end

  # Tools for managing tmp files in the hadoop dfs
  module Dfs
    module TmpFile
      def self.uuid
        @uuid ||= UUID.new
      end

      def self.tmp_ttl
        @tmp_ttl ||= (ENV['HADUPILS_TMP_TTL'] || '1209600').to_i
      end

      def self.tmp_path
        @tmp_path ||= (ENV['HADUPILS_TMP_PATH'] || '/tmp')
      end

      def self.tmpfile_path
        @tmpfile_path ||= ::File.join(tmp_path, "hadupils-tmp-#{uuid.generate(:compact)}")
      end

      def self.reset_tmpfile!
        @tmpdir_path = nil
      end
    end
  end

  # Tools for managing hadoop configuration files ("hadoop.xml").
  module HadoopConf
    module HadoopOpt
      def hadoop_opts
        ['-conf', path]
      end
    end

    # Wraps an extant hadoop configuration file and provides
    # an interface compatible with the critical parts of the
    # Static sibling class so they may be used interchangeably
    # by runners when determining hadoop options.
    class Static
      attr_reader :path

      include HadoopOpt

      # Given a path, expands it ti
      def initialize(path)
        @path = ::File.expand_path(path)
      end

      def close
      end
    end
  end

  # Tools for managing hive initialization files ("hiverc").
  module HiveRC
    module HiveOpt
      def hive_opts
        ['-i', path]
      end
    end

    # Wraps an extant hive initialization file, and providing
    # an interface compatible with the critical parts of the
    # Dynamic sibling class so they may be used interchangeably
    # by runners when determining hive options.
    class Static
      attr_reader :path

      include HiveOpt

      # Given a path, expands it ti
      def initialize(path)
        @path = ::File.expand_path(path)
      end

      def close
      end
    end

    # Manages dynamic hive initialization files, assembling a temporary file
    # and understanding how to write assets/lines into the initialization file for
    # use with hive's -i option.
    class Dynamic
      attr_reader :file

      include HiveOpt

      # This will allow us to change what handles the dynamic files.
      def self.file_handler=(handler)
        @file_handler = handler
      end

      # The class to use for creating the files; defaults to ::Tempfile
      def self.file_handler
        @file_handler || ::Tempfile
      end

      # Sets up a wrapped file, using the class' file_handler,
      def initialize
        @file = self.class.file_handler.new('hadupils-hiverc')
      end

      def path
        ::File.expand_path @file.path
      end

      def close
        @file.close
      end

      # Writes the items to the file, using #hiverc_command on each item that
      # responds to it (Hadupils::Assets::* instances) and #to_s on the rest.
      # Separates lines by newline, and provides a trailing newline.  However,
      # the items are responsible for ensuring the proper terminating semicolon.
      # The writes are flushed to the underlying file immediately afterward.
      def write(items)
        lines = items.collect do |item|
          if item.respond_to? :hiverc_command
            item.hiverc_command
          else
            item.to_s
          end
        end
        @file.write(lines.join("\n") + "\n")
        @file.flush
      end
    end
  end

  class EvalProxy
    def initialize(scope)
      @scope = scope
    end

    def assets(&block)
      @scope.instance_eval do
        @assets_block = block
      end
    end

    def hadoop_conf(&block)
      @scope.instance_eval do
        @hadoop_conf_block = block
      end
    end

    def hiverc(&block)
      @scope.instance_eval do
        @hiverc_block = block
      end
    end
  end

  class Base
    attr_reader :assets, :path

    def initialize(directory, &block)
      if block_given?
        EvalProxy.new(self).instance_eval &block
      end
      @path = ::File.expand_path(directory) unless directory.nil?
      @assets = merge_assets(self.class.gather_assets(@path))
    end

    def merge_assets(assets)
      return @assets_block.call(assets) if @assets_block
      assets
    end

    def hadoop_confs
      []
    end

    def hivercs
      []
    end

    def self.gather_assets(directory)
      if directory
        Hadupils::Assets.assets_in(directory)
      else
        []
      end
    end
  end

  class Flat < Base
    def hivercs
      @hiverc ||= assemble_hiverc
      [@hiverc]
    end

    def default_hiverc_items
      @assets.dup
    end

    def assemble_hiverc
      assets = default_hiverc_items
      if @hiverc_block
        assets = @hiverc_block.call(assets.dup)
      end
      hiverc = Hadupils::Extensions::HiveRC::Dynamic.new
      hiverc.write(assets)
      hiverc
    end
  end

  class FlatArchivePath < Flat
    def archives_for_path_env
      @assets.find_all do |a|
        if a.kind_of? Hadupils::Assets::Archive
          begin
            Hadupils::Util.archive_has_directory?(a.path, 'bin')
          rescue
            false
          end
        else
          false
        end
      end
    end

    def default_hiverc_items
      items = super
      archs = archives_for_path_env
      if archs.length > 0
        items << self.class.assemble_path_env(archs)
      end
      items
    end

    def self.assemble_path_env(archives)
      paths = archives.collect {|a| "$(pwd)/#{ a.name }/bin" }
      "SET mapred.child.env = PATH=#{ paths.join(':') }:$PATH;"
    end
  end

  class Static < Base
    def self.gather_assets(path)
      []
    end

    def hadoop_conf_path
      ::File.join(path, 'hadoop.xml') if path
    end

    def hiverc_path
      ::File.join(path, 'hiverc') if path
    end

    def hadoop_conf?
      hadoop_conf_path ? ::File.file?(hadoop_conf_path) : false
    end

    def hiverc?
      hiverc_path ? ::File.file?(hiverc_path) : false
    end

    def hadoop_confs
      r = []
      r << Hadupils::Extensions::HadoopConf::Static.new(hadoop_conf_path) if hadoop_conf?
      r
    end

    def hivercs
      r = []
      r << Hadupils::Extensions::HiveRC::Static.new(hiverc_path) if hiverc?
      r
    end
  end
end

require 'hadupils/extensions/hive'
